Sources/Dejavu/GRDBSession.swift

```diff
@@ -14,78 +14,87 @@
+            guard var instanceCount = instanceCounts[request.hashString] else {
+                return 0
             }
+            instanceCount -= 1
+            instanceCounts[request.hashString] = instanceCount
+            return instanceCount
         }
     }
     
-    func clearCache() {
-        GRDBSession.serialQueue.sync {
-            // if db file exists, then delete it
-            if FileManager.default.fileExists(atPath: configuration.fileURL.path) {
-                try? FileManager.default.removeItem(at: configuration.fileURL)
-            }
-        }
-    }
+    private let state = OSAllocatedUnfairLock(initialState: State())
+    
+    let dbQueue: DatabaseQueue
     
-    private func setupDBForRecording() {
-        guard dbQueue == nil else { return }
-        if configuration.mode == .cleanRecord {
-            // if db file exists, then delete it
-            if FileManager.default.fileExists(atPath: configuration.fileURL.path) {
-                try? FileManager.default.removeItem(at: configuration.fileURL)
+    required init(configuration: DejavuSessionConfiguration) throws {
+        log("Initializing Dejavu GRDB session: \(configuration.fileURL)", category: .beginSession, type: .info)
+        
+        lazy var temporaryURL = URL.temporaryDirectory
+            .appending(component: ProcessInfo().globallyUniqueString)
+        
+        switch configuration.mode {
+        case .cleanRecord:
+            dbQueue = try Self.makeDatabaseAndSchema(fileURL: temporaryURL)
+        case .supplementalRecord:
+            if FileManager.default.fileExists(at: configuration.fileURL) {
+                try FileManager.default.copyItem(
+                    at: configuration.fileURL,
+                    to: temporaryURL
+                )
+                dbQueue = try DatabaseQueue(fileURL: temporaryURL)
+            } else {
+                dbQueue = try Self.makeDatabaseAndSchema(fileURL: temporaryURL)
             }
-            
-            createDatabaseAndSchema()
-        } else if configuration.mode == .supplementalRecord {
-            if !FileManager.default.fileExists(atPath: configuration.fileURL.path) {
-                // if database doesn't exist, create it
-                createDatabaseAndSchema()
+        case .playback:
+            if FileManager.default.fileExists(at: configuration.fileURL) {
+                // Only try to open if the file exists. In playback mode, don't
+                // create a database if there isn't one.
+                dbQueue = try DatabaseQueue(fileURL: configuration.fileURL)
             } else {
-                // otherwise open it
-                dbQueue = try? DatabaseQueue(path: configuration.fileURL.path)
+                dbQueue = try Self.makeDatabaseAndSchema(fileURL: temporaryURL)
             }
         }
+        
+        self.configuration = configuration
+    }
+    
+    func clearCache() {
```

Comment:
I'm not seeing this function get called when I'm setting the mode to `cleanRecord` and new databases are not getting created.

Reply:
This function is not being called, either internally or externally, which is why we're deprecated it. One of the updates in this PR is to perform database operations in a temporary directory and then remove/replace the old database when the session ends. So for `cleanRecord`, that is now done in `end()`. However, [one of the URLs](https://github.com/Esri/Dejavu/blob/7301e9214d7c51f4b93b894bf5ba6610f7ce7767/Sources/Dejavu/GRDBSession.swift#L318) being used isn't correct, so I'll update the PR to fix that. Thanks for finding that!

Reply:
I tested cleanRecord and it worked for me. Is it still not working for you?